Signal/src/ViewControllers/MediaDetailViewController.m

```diff
@@ -0,0 +1,902 @@
+    // better to just have everythign "flit" in/out.
+    [[UIApplication sharedApplication] setStatusBarHidden:areToolbarsHidden withAnimation:UIStatusBarAnimationNone];
+    [self.navigationController setNavigationBarHidden:areToolbarsHidden animated:NO];
+    self.videoProgressBar.hidden = areToolbarsHidden;
+
+    [UIView animateWithDuration:0.1
+                     animations:^(void) {
+                         self.view.backgroundColor = areToolbarsHidden ? UIColor.blackColor : UIColor.whiteColor;
+                         self.footerBar.alpha = areToolbarsHidden ? 0 : 1;
+                     }];
+}
+
+- (void)initializeGestureRecognizers
+{
+    UITapGestureRecognizer *doubleTap =
+        [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didDoubleTapImage:)];
+    doubleTap.numberOfTapsRequired = 2;
+    [self.view addGestureRecognizer:doubleTap];
+
+    UITapGestureRecognizer *singleTap =
+        [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didTapImage:)];
+    [singleTap requireGestureRecognizerToFail:doubleTap];
+
+    [self.view addGestureRecognizer:singleTap];
+
+    // UISwipeGestureRecognizer supposedly supports multiple directions,
+    // but in practice it works better if you use a separate GR for each
+    // direction.
+    for (NSNumber *direction in @[
+                                  @(UISwipeGestureRecognizerDirectionRight),
+                                  @(UISwipeGestureRecognizerDirectionLeft),
+                                  @(UISwipeGestureRecognizerDirectionUp),
+                                  @(UISwipeGestureRecognizerDirectionDown),
+                                  ]) {
+        UISwipeGestureRecognizer *swipe =
+            [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(didSwipeImage:)];
+        swipe.direction = (UISwipeGestureRecognizerDirection) direction.integerValue;
+        swipe.delegate = self;
+        [self.view addGestureRecognizer:swipe];
+    }
+
+    UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc] initWithTarget:self
+                                                                                            action:@selector(longPressGesture:)];
+    longPress.delegate = self;
+    [self.view addGestureRecognizer:longPress];
+}
+
+#pragma mark - Gesture Recognizers
+
+- (void)didTapDismissButton:(id)sender
+{
+    [self dismissSelfAnimated:YES completion:nil];
+}
+
+- (void)didTapImage:(id)sender
+{
+    DDLogVerbose(@"%@ did tap image.", self.logTag);
+    self.areToolbarsHidden = !self.areToolbarsHidden;
+}
+
+- (void)didDoubleTapImage:(id)sender
+{
+    DDLogVerbose(@"%@ did tap image.", self.logTag);
+    if (self.scrollView.zoomScale == self.scrollView.minimumZoomScale) {
+        [self.scrollView setZoomScale:self.scrollView.minimumZoomScale * 2 animated:YES];
+    } else {
+        [self.scrollView setZoomScale:self.scrollView.minimumZoomScale animated:YES];
```

Comment:
Shouldn't double-tap zoom in on the gesture location?

Reply:
Yeah, great point. done!